File: testing/plot.py
import pandas as pd
import matplotlib.pyplot as plt
#  from mpl_toolkits.mplot3d import Axes3D

if __name__ == "__main__":

    # Read the CSV file
    df = pd.read_csv('output.csv')

    # Create a 3D plot
    fig = plt.figure(figsize=(10, 8))
    ax = fig.add_subplot(111, projection='3d')

    # Plot the x, y, z coordinates
    ax.scatter(df['x'], df['y'], df['z'], c=df['scalar'],
               cmap='viridis', s=0.1)
    #  ax.plot(df['x'], df['y'], df['z'], color='red', alpha=0.5)
    ax.set_xlim([-1000, 1000])
    ax.set_ylim([-1000, 1000])
    ax.set_zlim([0, 1000])

    # Add labels and title
    ax.set_xlabel('X')
    ax.set_ylabel('Y')
    ax.set_zlabel('Z')
    ax.set_title('3D Visualization of Quaternion Data')

    # Add a color bar to show scalar values
    plt.colorbar(ax.scatter(df['x'], df['y'], df['z'],
                 c=df['scalar'], cmap='viridis'))

    # Show the plot
    plt.tight_layout()
    plt.show()

File: testing/plot.jl
using Pkg

Pkg.add(["GLMakie", "CSV", "DataFrames"])

using GLMakie
using CSV
using DataFrames

function plot_quaternion_data(filename)
    # Read the CSV file
    df = CSV.read(filename, DataFrame)

    # Create the figure and axis
    fig = Figure(size=(1000, 800))
    ax = Axis3(fig[1, 1],
        xlabel="X",
        ylabel="Y",
        zlabel="Z",
        title="3D Visualization of Quaternion Data"
    )

    # Scatter plot with color mapping
    scatter!(ax,
        df.x, df.y, df.z,
        color=df.scalar,
        colormap=:viridis,
        markersize=0.1
    )

    # Set axis limits
    #  xlims!(ax, -1000, 1000)
    #  ylims!(ax, -1000, 1000)
    #  zlims!(ax, 0, 1000)

    # Add colorbar
    Colorbar(fig[1, 2], colormap=:viridis)

    # Display the figure
    display(fig)
end

# Usage
plot_quaternion_data("output.csv")
